flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/OneSlotPerExecutionSlotAllocator.java

```diff
@@ -0,0 +1,277 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * This slot allocator will request one physical slot for each single execution vertex.
+ * The slots will be requested in bulks so that the {@link SlotProvider} can check
+ * whether this bulk of slot requests can be fulfilled at the same time.
+ * It has several limitations:
+ *
+ * <p>1. Slot sharing will be ignored.
+ *
+ * <p>2. Co-location constraints are not allowed.
+ *
+ * <p>3. Intra-bulk input location preferences will be ignored.
+ */
+public class OneSlotPerExecutionSlotAllocator extends AbstractExecutionSlotAllocator {
+
+	private static final Logger LOG = LoggerFactory.getLogger(OneSlotPerExecutionSlotAllocator.class);
+
+	private final SlotProvider slotProvider;
+
+	private final SlotOwner slotOwner;
+
+	private final boolean slotWillBeOccupiedIndefinitely;
+
+	private final Time allocationTimeout;
+
+	public OneSlotPerExecutionSlotAllocator(
+			final SlotProvider slotProvider,
+			final InputsLocationsRetriever inputsLocationsRetriever,
+			final boolean slotWillBeOccupiedIndefinitely,
+			final Time allocationTimeout) {
+
+		super(inputsLocationsRetriever);
+
+		this.slotProvider = checkNotNull(slotProvider);
+		this.slotWillBeOccupiedIndefinitely = slotWillBeOccupiedIndefinitely;
+		this.allocationTimeout = checkNotNull(allocationTimeout);
+
+		this.slotOwner = new OneSlotPerExecutionSlotAllocatorSlotOwner();
+	}
+
+	@Override
+	public List<SlotExecutionVertexAssignment> allocateSlotsFor(
+			final List<ExecutionVertexSchedulingRequirements> executionVertexSchedulingRequirements) {
+
+		validateSchedulingRequirements(executionVertexSchedulingRequirements);
+
+		validateNoCoLocationConstraint(executionVertexSchedulingRequirements);
+
+		final Set<ExecutionVertexID> allExecutionVertexIds = executionVertexSchedulingRequirements.stream()
+			.map(ExecutionVertexSchedulingRequirements::getExecutionVertexId)
+			.collect(Collectors.toSet());
+
+		final Map<ExecutionVertexID, SlotRequestId> executionVertexSlotRequestIds =
+			generateExecutionVertexSlotRequestIds(allExecutionVertexIds);
+
+		final List<SlotExecutionVertexAssignment> slotExecutionVertexAssignments =
+			createSlotExecutionVertexAssignments(executionVertexSchedulingRequirements, executionVertexSlotRequestIds);
+
+		final List<CompletableFuture<PhysicalSlotRequest>> physicalSlotRequestFutures =
+			createPhysicalSlotRequestFutures(
+				executionVertexSchedulingRequirements,
+				allExecutionVertexIds,
+				executionVertexSlotRequestIds);
+
+		allocateSlotsForAssignments(
+			physicalSlotRequestFutures,
+			slotExecutionVertexAssignments,
+			executionVertexSlotRequestIds);
+
+		return slotExecutionVertexAssignments;
```

Comment:
Good suggestion!

I had wanted but failed to do it due to the `slotRequestFailureHandler` parameter of `createAndRegisterSlotExecutionVertexAssignment` was not well designed so that slotRequestId was needed before creating `SlotExecutionVertexAssignment`s. Just changed it in b0368079607316f7bc0791609c7b2128750740c3.

This method is now reorganized regarding your suggestions with some small differences:
1. added `slotRequestId` to `SlotExecutionVertexAssignment` and removed `generateExecutionVertexSlotRequestIds `
2. `allExecutionVertexIds` is moved into `createPhysicalSlotRequestFutures(...)`
3. `allocateSlotsForAssignments` is simplified
4. statements are reordered

Reply:
Are you worried about performance for `new ArrayList<>(slotExecutionVertexAssignments.values())` to make `slotExecutionVertexAssignments` a map?
Then you would not need `allExecutionVertexIds` in `createPhysicalSlotRequestFutures` as it could be just `slotExecutionVertexAssignments.keySet()`.

Also not sure, why the API is to return `List<SlotExecutionVertexAssignment>`. Do we need list indexing anywhere on the using side? Otherwise, `allocateSlotsFor` could return `Collection`, then it could return just `slotExecutionVertexAssignments.values()`.